### PR TITLE
Discriminator Groups

### PR DESCRIPTION
**Pull request by maennchen**
Clone of schmittjoh/serializer#567

### CODE REVIEW

**Comment by maennchen** [comment]
@schmittjoh:
> would you mind re-basing the discriminator commit and adjusting the serialize/unserialize method so that it is BC by either throwing an exception or setting a default value?

I don't understand exactly what you want here. Could you explain further?

**Comment by schmittjoh** [comment]
Just a check to handle unserializing an old version of the metadata class that didn't have the property yet. Right now, we will drop the parent metadata or have a notice error I guess.

File: src/JMS/Serializer/Metadata/ClassMetadata.php [hunk 1]
```diff
@@ -251,6 +255,7 @@ public function unserialize($str)
             $this->discriminatorFieldName,
             $this->discriminatorValue,
             $this->discriminatorMap,
+            $this->discriminatorGroups,
             $parentStr
         ) = unserialize($str);
```
Comment (schmittjoh) [inline src/JMS/Serializer/Metadata/ClassMetadata.php]: ``$str`` might not have data for the new property yet.

**Comment by maennchen** [comment]
Do we have to provide no BC breaks for Caches? I think that people have to clear caches between deployments.

**Comment by schmittjoh** [comment]
I haven't looked into the consequence, but if we for example do not pick-up some exclusion rules anymore because of such a change and someone forgets to clear his cache, sensitive data might be exposed.

So, in general, yes, we should not break BC and anyway it's easy for us to add this check here.

**Comment by wins1908** [comment]
Hello @schmittjoh , @maennchen 
Could we continue on this pull request?

**Comment by maennchen** [comment]
@schmittjoh: 
I added a check for consistency.

I'm not really happy about the solution. (Or that we have to check caches in general.) What do you think about it?

File: src/JMS/Serializer/Metadata/ClassMetadata.php [hunk 2]
```diff
@@ -34,6 +34,7 @@ class ClassMetadata extends MergeableClassMetadata
     const ACCESSOR_ORDER_UNDEFINED = 'undefined';
     const ACCESSOR_ORDER_ALPHABETICAL = 'alphabetical';
     const ACCESSOR_ORDER_CUSTOM = 'custom';
+    const COUNT_PARAMS = 16;
```
Comment (goetas) [inline src/JMS/Serializer/Metadata/ClassMetadata.php]: why do not use this approach https://github.com/schmittjoh/serializer/pull/301/files#diff-2d7dc6b1d453d596671e68f15d9a36d0R158 to have a backward compatibile serializtion ? and does not require all che "count" checks

Comment (maennchen) [inline src/JMS/Serializer/Metadata/ClassMetadata.php]: I do not want backward compatible serialization. People have to clear caches between deployments.

Comment (maennchen) [inline src/JMS/Serializer/Metadata/ClassMetadata.php]: In my opinion caches are not part of backwards compatibility. One can expect the users to clean caches.

**Comment by goetas** [comment]
> I do not want backward compatible serialization. People have to clear caches between deployments.

I do not see nothing wrong with keeping compatibility, and the proposed solution requires less code and logic.

**Comment by maennchen** [comment]
I think having compatibility to caches is nonsense.

I'm not going to contribute to OS projects with code that does not live up to the standards I set for myself.

If you want the method with some other deserialization handling you're going to have to commit it yourself.

**Pull request merged**